scripts/builtin/xgboost.dml

```diff
@@ -0,0 +1,780 @@
+# INPUT:    residuals_vector: a 1xn matrix (one feature with all residuals)
+# OUTPUT:   similarity_score: the similarity score of the residuals
+calculateOutputValue = function (matrix[Double] residuals_vector, Double lambda) return (Double output_value) {
+  output_value = (sum(residuals_vector)) / (nrow(residuals_vector) + lambda);
+  if(output_value == 'NaN') # just in case we have a node with no sample inside
+    output_value = 0.0
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    residuals_vector: a 1xn matrix (one feature with all residuals)
+# OUTPUT:   similarity_score: the similarity score of the residuals
+calculateOutputValueClassification = function (matrix[Double] residuals_vector, matrix[Double] predictions, Double lambda) return (Double output_value) {
+    nominator = (sum(residuals_vector))
+    d =  predictions * (1 - predictions)
+    denominator = sum(d) + lambda
+    if(denominator == 0)
+      output_value = 0
+    else
+      output_value = nominator / denominator
+}
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    row_vector: a 1xn matrix (one feature with all rows)
+#           prediction: a 1xn matrix, the current prediction value of each row
+# OUTPUT:   the residuals from the "prediction" as 1xn matrix
+calculateResiduals = function (matrix[Double] row_vector, matrix[Double] prediction) return (matrix[Double] residuals_row) {
+   assert(ncol(row_vector) == 1)
+   residuals_row = row_vector - prediction;
+}
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    row_vector: a 1xn matrix which has unordered values
+# OUTPUT:   row_vector_ordered: 1xn matrix were the values are in accessing order
+orderFeature = function (matrix[double] row_vector) return (matrix[double] row_vector_ordered) {
+    row_vector_ordered = order(target=row_vector, by=1, decreasing=FALSE, index.return=FALSE);
+}
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    m: a nxn matrix which should be set to NaN's in each cell
+# OUTPUT:   new_m: matrix with same dimensions with all NaNs inside
+fillMatrixWithNaN = function(Matrix[Double] m) return (Matrix[Double] new_m) {
+  new_m = matrix(0, rows=nrow(m), cols=ncol(m))
+  for(i in 1:ncol(m))
+  {
+    for(j in 1:nrow(m))
+      new_m[i,j] = 'NaN'
+  }
+}
+
+
+average = function(Matrix[Double] x, Matrix[Double] y) return (Double avg) {
+  assert(ncol(x) == 1 & nrow(x) == 1)
+  assert(ncol(y) == 1 & nrow(y) == 1)
+  avg = (as.scalar(x) + as.scalar(y)) / 2
+}
+
+
+
+# -------------------------------------------------------------------------------------------
+# DEBUG
+#
+printMatrix = function(Matrix[Double] m, string name) return() {
```

Comment:
remove these debug functions once you don't need them anymore.

Reply:
deleted in https://github.com/apache/systemds/pull/1334/commits/dc3a9f489951d7e13ec47c5181d2c5d7022665ce